### PR TITLE
MLJTuning PR 75 (API Changes)

### PR DESCRIPTION
**Pull request by yalwan-iqvia**
Thank you for your contribution. You're a :star: already!

Draft PR from @ablaom for upcoming MLJTuning API changes (opened by @yalwan-iqvia)

## Checklist

- [x] Your code builds clean without any errors or warnings
- [x] Unit tests have been added for core changes
- [x] `julia --project -e 'using Pkg; Pkg.test()'` has been run locally and passes
- [ ] Documentation has been updated with corresponding changes (if applicable)

### CODE REVIEW

**Comment by ablaom** [comment]
Requires: 

- [ ] https://github.com/alan-turing-institute/MLJTuning.jl/pull/75

File: src/MLJTreeParzen.jl [hunk 1]
```diff
@@ -242,24 +245,35 @@ function MLJTuning.models!(
...
 
     # when num_hist == 0 this is the full number of random (independent draws) (minus suggestions as drawn above)
     # else its the max number of times we can ask TP from its distributions
     for i in 1:max_draws
-        push!(candidates, ask(state.space, trialhist, strategy.config))
+        push!(candidates, ask(space.space, trialhist, strategy.config))
     end
 
-    return [
+    newstate = (space=space, trialhist=trialhist)
```
Comment (yalwan-iqvia) [inline src/MLJTreeParzen.jl:257]: I spent a while (I mean just a few minutes) trying to grok why we could eat things from the state in the way we were, but I see now with this more functional approach to `models` we return a constructed state: I approve

File: src/MLJTreeParzen.jl [hunk 2]
```diff
@@ -198,40 +198,43 @@ end
...
+    space = state.space
+    trialhist = state.trialhist
+
+    num_hist = length(vector(history))
 
-    num_hist = histlen(history)
-    num_suggest = num_hist == 0 ? length(state.suggestions) : 0
+    # get an up-to-date the history of trial objects by appending to
+    # the trial object history stored in `state`:
+    Δhistory = view(vector(history), (length(trialhist) + 1):num_hist) # recent MLJ history
```
Comment (yalwan-iqvia) [inline src/MLJTreeParzen.jl:230]: I get (and appreciate) what you're doing here, but I'm not a fan of symbols in code -- please stick to alnums & underscores

File: test/MLJ/unit.jl [hunk 3]
```diff
@@ -92,9 +110,9 @@ end
         suggestions = Dict{Symbol}[]
         tuning, state = setup(;n_startup=3,space=space, suggest=suggestions)
 
-        output = MLJTuning.models!(tuning, testmodel, nothing, state, 0, 0)
+        output, _ = MLJTuning.models(tuning, testmodel, nothing, state, 0, 0)
 
-        @test output isa Vector{Tuple{Any, TreeParzen.Trials.Trial}}
+        @test output isa Vector{<:Tuple{Any, TreeParzen.Trials.Trial}}
```
Comment (yalwan-iqvia) [inline test/MLJ/unit.jl:115]: How come this was necessary?

File: test/MLJ/unit.jl [hunk 4]
```diff
@@ -141,15 +159,27 @@ end
...
 
-                output = MLJTuning.models!(tuning, testmodel, fakehist[1:histlen], state, 0, 0)
+                output, state = MLJTuning.models(tuning,
+                                           testmodel,
+                                           fakeMLJhist[1:histlen],
+                                           state,
+                                           0,
+                                           0)
                 @test length(output) == sim_value
+                @test length(state.trialhist) == histlen
```
Comment (yalwan-iqvia) [inline test/MLJ/unit.jl:182]: nice 👍

File: src/MLJTreeParzen.jl [hunk 5]
```diff
@@ -198,40 +198,43 @@ end
...
 MLJTuning.default_n(tuning::MLJTreeParzenTuning, range) = 50
 
+vector(history::Nothing) = Trials.Trial[]
+vector(history::Vector) = history
+
+get_trialhist(history) =
+    map(history) do entry
+        trial_object = entry.metadata
+        measurement = entry.measurement[1]
+        # @ablaom asks "Is this deepcopy really necessary?":
```
Comment (yalwan-iqvia) [inline src/MLJTreeParzen.jl:208]: I'm not sure its "strictly" necessary, however, I didn't really want to reason too much about mutated variables and references and so on, so would rather to copy the incomplete object and return the completed copy

File: test/MLJ/unit.jl (hunk 3 above)
Comment (ablaom) [reply test/MLJ/unit.jl:115]: It was failing without the change. Julia's type system is not ["covariant"](https://docs.julialang.org/en/v1/manual/types/). So for example:

```julia
julia> [(4, 3), ] isa Vector{Tuple{Any, Any}}
false

julia> [(4, 3), ] isa Vector{<:Tuple{Any, Any}}
true
```
But I guess you probably knew that. I admit that I don't understand is why the old tests were passing.

**Comment by ablaom** [comment]
@yalwan-iqvia I'm all done if you're happy with my responses.

**Comment by yalwan-iqvia** [comment]
@iqml as I have "opened" this PR, github wouldn't allow me to approve -- could you approve?

**Comment by yalwan-iqvia** [comment]
@ablaom Thanks for taking care of this and quickly at that. I hope you didn't find this piece of work too much trouble! I think we have no upcoming planned features for now for this project as well, so it would warrant an immediate release of TreeParzen 0.2 (could you add that into Project.toml?)

**Review by iqml — APPROVED** [review]

**Comment by yalwan-iqvia** [comment]
Thanks for this Anthony! I'll merge now and update documentation before I do a final release

**Pull request merged**